packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,406 @@
+		global $wpdb;
+		$table_name = self::get_table_name();
+
+		// Drop the table.
+		$wpdb->query( $wpdb->prepare( 'DROP TABLE IF EXISTS %s', $table_name ) );
+
+		// Delete the option.
+		delete_option( self::OPTION_NAME );
+	}
+
+	/**
+	 * Reset the class.
+	 */
+	public function reset() {
+		$this->set_session( null );
+	}
+
+	/**
+	 * Set the current session ID.
+	 *
+	 * @param int|null $session_id The session ID.
+	 */
+	public function set_session( $session_id ) {
+		$this->current_session = $session_id;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id = null ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ?? $this->current_session ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Map an entity to the index. If $id is provided, it will be used to map the entity.
+	 *
+	 * @param string   $entity_type The type of the entity.
+	 * @param array    $data The data to map.
+	 * @param int|null $id The ID of the entity.
+	 * @param int|null $additional_id The additional ID of the entity.
+	 */
+	public function map_entity( $entity_type, $data, $id = null, $additional_id = null ) {
+		global $wpdb;
+
+		if ( ! array_key_exists( $entity_type, self::ENTITY_TYPES ) ) {
+			return;
+		}
+
+		$new_entity = array(
+			'session_id'   => $this->current_session,
+			'entity_type' => self::ENTITY_TYPES[ $entity_type ],
+			'entity_id'   => null,
+			'mapped_id'    => is_null( $id ) ? null : (string) $id,
+			'parent_id'    => null,
+			'byte_offset'  => 0,
+			// Items with a parent has at least a sort order of 2.
+			'sort_order'   => 1,
+		);
+		// Get the ID of the entity.
+		$entity_id = (string) $data[ self::ENTITY_TYPES_ID[ $entity_type ] ];
+
+		// Map the parent ID if the entity has one.
+		switch ( $entity_type ) {
+			// @TODO: missing comment parent ID.
+			case 'comment_meta':
+				if ( array_key_exists( 'comment_id', $data ) ) {
+					$new_entity['parent_id'] = $data['comment_id'];
+				}
+				break;
+			case 'post':
+				if ( 'post' === $data['post_type'] || 'page' === $data['post_type'] ) {
+					if ( array_key_exists( 'post_parent', $data ) && '0' !== $data['post_parent'] ) {
+						$new_entity['parent_id'] = $data['post_parent'];
+					}
+				}
+				break;
+			case 'post_meta':
+				if ( array_key_exists( 'post_id', $data ) ) {
+					$new_entity['parent_id'] = $data['post_id'];
+				}
+				break;
+			case 'term':
+				if ( array_key_exists( 'parent', $data ) ) {
+					$new_entity['parent_id'] = $data['parent'];
+				}
+				break;
+			case 'term_meta':
+				if ( array_key_exists( 'term_id', $data ) ) {
+					$new_entity['parent_id'] = $data['term_id'];
+				}
+				break;
+		}
+
+		// The entity has been imported, so we can use the ID.
+		if ( $id ) {
```

Comment:
What's the rationale for calling a topological sorter that is meant to sort the entity stream before we start importing,  after an entity is imported?

Reply:
Or is it doubling as an ID mapper? I'm not sure if topological sorter is the best place for that and it might take us some discussing to find the right place. Let's remove this part for now and focus just on pre-import sorting.